### ISSUE TITLE
TimetableController's scrollController does not update its page value

### ISSUE THREAD

**Issue by RedMarbles** [JonasWanke/timetable#19]
**Description**

Scrolling on the calendar doesn't seem to change the value of the current page. I added the following listener:

```dart
timetableController.scrollControllers.addPageChangedListener(() {
      print('Page changed to ${timetableController.scrollControllers.page}');
    });
```

This causes it to print a message every time I try to scroll, but the value of the page printed always remains the same.

As a side-effect of this, other listeners that depend on the page value, such as `TimetableController.dateListenable` and `TimetableController.currentlyVisibleDatesListenable` are also triggering their listeners, but their values aren't changing, and are remaining the same as the `initialDate` and `initialTimeRange` values.

**Screenshots**

![image](https://user-images.githubusercontent.com/5616344/85153369-6afb3100-b256-11ea-8fbb-0f624239f7de.png)

**Environment:**

- Device:  Huawei Honor 9 Lite
- OS version:  Android 9.0.0
- Package version:  0.2.3

**Comment by JonasWanke** [comment]
Thanks for reporting that issue and sorry for the delay, I hope I can take a detailed look at it today or tomorrow.

**Comment by JonasWanke** [comment]
Unfortunately, I cannot reproduce your issue. Using the example app included in release v0.2.3 and adding your code snippet, I get the following log output:

```
I/flutter (28892): Page changed to 18438.0
I/flutter (28892): Page changed to 18438.022012578615
I/flutter (28892): Page changed to 18438.040880503144
I/flutter (28892): Page changed to 18438.059748427677
I/flutter (28892): Page changed to 18438.075471698117
I/flutter (28892): Page changed to 18438.08805031447
I/flutter (28892): Page changed to 18438.10062893082
I/flutter (28892): Page changed to 18438.113207547172
I/flutter (28892): Page changed to 18438.125786163524
I/flutter (28892): Page changed to 18438.138364779876
I/flutter (28892): Page changed to 18438.150943396227
I/flutter (28892): Page changed to 18438.16037735849
I/flutter (28892): Page changed to 18438.169811320753
I/flutter (28892): Page changed to 18438.17924528302
I/flutter (28892): Page changed to 18438.19182389937
I/flutter (28892): Page changed to 18438.204402515723
I/flutter (28892): Page changed to 18438.213836477986
I/flutter (28892): Page changed to 18438.226415094337
I/flutter (28892): Page changed to 18438.23899371069
I/flutter (28892): Page changed to 18438.248427672952
I/flutter (28892): Page changed to 18438.261006289304
I/flutter (28892): Page changed to 18438.270440251566
...
```

Just to confirm, that's the same controller you assign to the `Timetable` widget? And are able to scroll the widget horizontally while reading those values, or doesn't it move at all while dragging?

**Comment by RedMarbles** [comment]
That's right, it's the same controller. And I'm able to scroll the widget horizontally without a problem, 
However, the small bar that indicates the current time doesn't move when I scroll. Which means that even when I scroll to a different date, the bar stays at the same location.
I don't know if this is related to the same bug, but I'm also unable to use the `timetableController.animateTo()` function to scroll to a different date. The calendar remains in the same position without scrolling when I call the `animateTo()` method.

**Comment by JonasWanke** [comment]
Interesting; I have no idea what's going on there. Could you please share the full code to reproduce this behavior?

**Comment by RedMarbles** [comment]
I tried to make a minimum viable example to share here, but noticed that it seems to work fine. I think it might be a bug on my side. I'll try to debug my side of the code before sending you another update.
Thanks again!

**Comment by RedMarbles** [comment]
I think I've found the problem. I need to make a network request to fetch the events when I reach the screen with the calendar, and I also need to keep re-fetching the data every minute or so. I've simulated this using a delayed Future in this example.

This problem with the unchanging page seems to happen when I call setState() to change the controller. The bug seems to be when the Timetable widget is rebuilt after I change the controller using setState() to provide it an updated set of Events.

This is definitely reloading the Timetable widget with the new set of events, but the controller no longer seems to be fully functional.

```dart
import 'package:flutter/material.dart';
import 'package:timetable/timetable.dart';
import 'package:time_machine/time_machine.dart';

class CalendarTest extends StatefulWidget {
  @override
  _CalendarTestState createState() => _CalendarTestState();
}

class _CalendarTestState extends State<CalendarTest> {
  TimetableController<BasicEvent> _controller;

  @override
  void initState() {
    super.initState();

    _fillController([]); // Calling the controller with an empty list of events

    // simulating fetching data from server after a delay
    Future.delayed(Duration(seconds: 5), () {
      setState(() {
        _fillController(_getEvents());
      });
    });
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  List<BasicEvent> _getEvents() {
    return <BasicEvent>[
      BasicEvent(
        id: 0,
        title: 'My Event',
        color: Colors.blue,
        start: LocalDate.today().at(LocalTime(13, 0, 0)),
        end: LocalDate.today().at(LocalTime(15, 0, 0)),
      ),
    ];
  }

  void _fillController(List<BasicEvent> events) {
    _controller = TimetableController(
      eventProvider: EventProvider.list(events),
    );
    _controller.scrollControllers.addPageChangedListener(() {
      // This stops working after the controller is changed
      print('Controller ${_controller.hashCode} - '
          'Page changed to ${_controller.scrollControllers.page}');
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Timetable<BasicEvent>(
        controller: _controller,
        eventBuilder: (event) => BasicEventWidget(event),
        allDayEventBuilder: (context, event, info) =>
            BasicAllDayEventWidget(event, info: info),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () {
          // This stops working after the controller is changed
          _controller.animateTo(LocalDate.today().addDays(-5));
        },
      ),
    );
  }
}
```

**Comment by JonasWanke** [comment]
Okay, I found the problem: Timetable manages state via the controller and doesn't support switching that. The intended way to include dynamically changing events is to use [`EventProvider.stream()`](https://pub.dev/documentation/timetable/latest/timetable/EventProvider/EventProvider.stream.html) and provide a stream of events. Your code could be refactored to the following, which works as intended:

```dart
class _CalendarTestState extends State<CalendarTest> {
  TimetableController<BasicEvent> _controller;
  // Create a [StreamController] for all events. in the `EventProvider` below,
  // create a new subscription every time the `currentlyVisibleDates` changes,
  // hence we need a broadcast stream.
  final _events = StreamController<List<BasicEvent>>.broadcast()..add([]);

  @override
  void initState() {
    super.initState();

    // Only create a single controller and keep that throughout the life of your
    // widget.
    _controller = TimetableController(
      eventProvider: EventProvider.stream(
        // [eventGetter] will be called whenever Timetable's visible range
        // changes. Just filter all of your events by that interval via the
        // extension function `List<Event>.intersectingInterval()`.
        eventGetter: (DateInterval dates) => _events.stream.map((e) {
          return e.intersectingInterval(dates);
        }),
      ),
    );
    _controller.scrollControllers.addPageChangedListener(() {
      // This stops working after the controller is changed
      print('Controller ${_controller.hashCode} - '
          'Page changed to ${_controller.scrollControllers.page}');
    });

    Future.delayed(Duration(seconds: 5), () {
      // No need for `setState()`: Timetable listens to the stream.
      _events.add(_getEvents());
    });
  }

  @override
  void dispose() {
    _controller.dispose();
    // Don't forget to close the `StreamController`.
    _events.close();
    super.dispose();
  }

  List<BasicEvent> _getEvents() { /* ... */ }

  @override
  Widget build(BuildContext context) { /* ... */ }
}
```

I can add a note to the README to make it clearer that changing the controller isn't supported.

**Comment by RedMarbles** [comment]
Right, that works perfectly. Thanks a lot!
I think it would be useful to mention the usage of the stream controller in the readme as well.

**Comment by JonasWanke** [comment]
I've updated the documentation and also added a new `EventProvider`: `EventProvider.simpleStream(eventStream)`. It basically implements the boilerplate of the above implementation, meaning you can now create a non-broadcast `StreamController` and use `EventProvider.simpleStream(controller.stream)` instead of manually filtering the relevant events for a given date range.

I'll release it when #20 is done as well, probably by tomorrow evening.

**Comment by JonasWanke** [comment]
Update: The documentation update & new `EventProvider` are now published as part of [v0.2.4](https://github.com/JonasWanke/timetable/releases/tag/v0.2.4).
